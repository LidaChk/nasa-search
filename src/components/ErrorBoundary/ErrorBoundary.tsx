import { Component, ErrorInfo, ReactNode } from 'react';

interface Props {
  children: ReactNode;
}

interface State {
  hasError: boolean;
  error: Error | null;
}

class ErrorBoundary extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      hasError: false,
      error: null,
    };
  }

  static getDerivedStateFromError(error: Error): State {
    return {
      hasError: true,
      error,
    };
  }

  componentDidCatch(error: Error, errorInfo: ErrorInfo): void {
    console.error('Error:', error);
    console.error('Error Info:', errorInfo);
  }

  render() {
    if (this.state.hasError) {
      return (
        <div className="error-boundary error-container">
          <div role="img" aria-hidden="true" className="error-emoji strong">
            (^-^*)
          </div>
          <div>Something went wrong</div>
          <div className="error-message">{this.state.error?.message}</div>
          <button onClick={() => this.setState({ hasError: false })}>
            Try Again
          </button>
        </div>
      );
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
