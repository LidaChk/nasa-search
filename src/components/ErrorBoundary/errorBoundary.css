.error-boundary {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  gap: 1rem;
  height: 100vh;
  text-align: center;
  background-color: var(--background-color);
  color: var(--text-color);
  font-family: var(--font-family);
  padding: 20px;
}

.error-emoji {
  font-size: 3rem;
}

.error-message {
  color: var(--accent-color);
}
