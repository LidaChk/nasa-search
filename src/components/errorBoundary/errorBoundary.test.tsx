import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import ErrorBoundary from './errorBoundary';

const ErrorComponent = (): React.JSX.Element => {
  throw new Error('Test error');
};

const NormalComponent = (): React.JSX.Element => <div>Normal component</div>;

const originalError = console.error;
beforeAll(() => {
  console.error = jest.fn();
});

afterAll(() => {
  console.error = originalError;
});

describe('ErrorBoundary', () => {
  it('renders children when there is no error', () => {
    render(
      <ErrorBoundary>
        <NormalComponent />
      </ErrorBoundary>
    );

    expect(screen.getByText('Normal component')).toBeInTheDocument();
  });

  it('renders error UI when there is an error', () => {
    render(
      <ErrorBoundary>
        <ErrorComponent />
      </ErrorBoundary>
    );

    expect(screen.getByText('Something went wrong')).toBeInTheDocument();
    expect(screen.getByText('Test error')).toBeInTheDocument();
    expect(screen.getByText('(^-^*)')).toBeInTheDocument();
    expect(screen.getByRole('button')).toHaveTextContent('Try Again');
  });

  it('resets error state when "Try Again" is clicked', () => {
    const { rerender } = render(
      <ErrorBoundary>
        <ErrorComponent />
      </ErrorBoundary>
    );

    const tryAgainButton = screen.getByRole('button', { name: /try again/i });
    rerender(
      <ErrorBoundary>
        <NormalComponent />
      </ErrorBoundary>
    );

    fireEvent.click(tryAgainButton);

    expect(screen.getByText('Normal component')).toBeInTheDocument();
  });

  it('has correct CSS classes', () => {
    render(
      <ErrorBoundary>
        <ErrorComponent />
      </ErrorBoundary>
    );

    const errorContainer = screen.getByText(
      'Something went wrong'
    ).parentElement;
    expect(errorContainer).toHaveClass('error-boundary', 'error-container');
  });

  it('calls componentDidCatch when error occurs', () => {
    const spy = jest.spyOn(ErrorBoundary.prototype, 'componentDidCatch');

    render(
      <ErrorBoundary>
        <ErrorComponent />
      </ErrorBoundary>
    );

    expect(spy).toHaveBeenCalled();
    spy.mockRestore();
  });

  it('matches error state snapshot', () => {
    const { container } = render(
      <ErrorBoundary>
        <ErrorComponent />
      </ErrorBoundary>
    );

    expect(container).toMatchSnapshot();
  });
});
