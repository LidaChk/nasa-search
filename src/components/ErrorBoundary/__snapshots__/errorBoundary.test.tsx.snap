// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary matches error state snapshot 1`] = `
<div>
  <div
    class="error-boundary error-container"
  >
    <div
      aria-hidden="true"
      class="error-emoji strong"
      role="img"
    >
      (^-^*)
    </div>
    <div>
      Something went wrong
    </div>
    <div
      class="error-message"
    >
      Test error
    </div>
    <button>
      Try Again
    </button>
  </div>
</div>
`;
